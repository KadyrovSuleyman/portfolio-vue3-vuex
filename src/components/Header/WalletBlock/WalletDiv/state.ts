import MODAL from '@/store/modal/types';
import { Store } from 'vuex';

export interface StateT {
  address: string,
  balance: string,
  coinName: string,
  coinLink: string,
  coinAbbreviation: string,

  showModal: () => Promise<any>,
}

export const adaptState = (store: Store<any>): StateT => ({
  address: store.state.account.address,
  balance: store.getters['account/balance'],
  coinName: store.getters['coin/name'],
  coinLink: store.getters['coin/icon'],
  coinAbbreviation: store.getters['coin/abbreviation'],

  showModal: () => store.dispatch('modal/show', MODAL.copied),
});
