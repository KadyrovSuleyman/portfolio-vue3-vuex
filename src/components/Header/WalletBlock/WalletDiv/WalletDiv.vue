<script setup lang="ts">

import { computed } from 'vue';
import Div from '@/elements/Div/Div.vue';
import Button from '@/elements/Button/Button.vue';
import Span from '@/elements/Span/Span.vue';
import propsObj from '@/elements/propsObj';
import { useStore } from 'vuex';
import AddressBlock from './AddressBlock/AddressBlock.vue';
import { adaptState } from './state';
import reductAddress from './logic';
import createCopyClickHandler from './handlers';

// eslint-disable-next-line no-undef
const props = defineProps({ ...propsObj });
const comp = computed(() => ({ elem: props.elem || 'walletDiv' }));

const store = useStore();
const state = computed(() => adaptState(store));

const copyClickHandler = createCopyClickHandler(state);

</script>

<template>
  <Div :block="props.block" :elem="comp.elem" :mods="props.mods" >
    <AddressBlock :block="comp.elem" :onClick="copyClickHandler">
      {{ reductAddress(state.address) }}
    </AddressBlock>
    <Span :block="comp.elem">
      {{ state.balance }} {{ state.coinAbbreviation }}
    </Span>

    <Button :block="comp.elem" :icon="'coin'">
      <img :src="state.coinLink" :alt="state.coinName">
    </Button>

    <Button :block="comp.elem" :icon="'chevron'"/>
  </Div>
</template>

<style lang="scss">
  @use 'styles' as *;
</style>
