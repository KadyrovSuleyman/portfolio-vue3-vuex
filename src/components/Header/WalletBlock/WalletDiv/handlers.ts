import { ComputedRef } from 'vue';
import { StateT } from './state';

const createCopyClickHandler = (state: ComputedRef<StateT>) => async () => {
  await navigator.clipboard.writeText(state.value.address);
  await state.value.showModal();
};

export default createCopyClickHandler;
