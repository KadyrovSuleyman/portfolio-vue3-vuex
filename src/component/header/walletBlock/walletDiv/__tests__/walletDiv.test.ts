import { mount, VueWrapper } from '@vue/test-utils';
import { createStore } from 'vuex';
import WalletDiv from '../walletDiv.vue';

jest.mock('../adapter');
jest.mock('../logic');
jest.mock('../handlers');

let wrapper: VueWrapper<any>;
afterEach(() => {
  wrapper.unmount();
});

it('connectBlock renders', () => {
  wrapper = mount(WalletDiv, { global: { plugins: [createStore({})] } });
  expect(wrapper.element.outerHTML).toMatchSnapshot();
});

it('watchs props changes', async () => {
  wrapper = mount(WalletDiv, { global: { plugins: [createStore({})] } });
  expect(wrapper.find('.walletDiv').classes()).toEqual(['walletDiv']);

  await wrapper.setProps({
    mods: {
      selected: true,
    },
  });
  expect(wrapper.find('.walletDiv').classes()).toEqual(['walletDiv', 'walletDiv__selected']);

  await wrapper.setProps({
    mods: {
      selected: false,
    },
  });
  expect(wrapper.find('.walletDiv').classes()).toEqual(['walletDiv']);
});

describe('outer store', () => {
  it('changeAddress -> changeBalance', async () => {
    const store = createStore({
      state: {
        address: 'first-address',
        balance: 'first-balance',
      },
      mutations: {
        changeAddress: (state) => { state.address = 'second-address'; },
        changeBalance: (state) => { state.balance = 'second-balance'; },
      },
    });

    wrapper = mount(WalletDiv, {
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.find('.addressBlock-span').text()).toBe('first-address');
    expect(wrapper.find('.walletDiv-span').text()).toBe('first-balance');

    store.commit('changeAddress');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.addressBlock-span').text()).toBe('second-address');
    expect(wrapper.find('.walletDiv-span').text()).toBe('first-balance');

    store.commit('changeBalance');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.addressBlock-span').text()).toBe('second-address');
    expect(wrapper.find('.walletDiv-span').text()).toBe('second-balance');
  });

  it('changeBalance -> changeAddress', async () => {
    const store = createStore({
      state: {
        address: 'first-address',
        balance: 'first-balance',
      },
      mutations: {
        changeAddress: (state) => { state.address = 'second-address'; },
        changeBalance: (state) => { state.balance = 'second-balance'; },
      },
    });

    wrapper = mount(WalletDiv, {
      global: {
        plugins: [store],
      },
    });

    expect(wrapper.find('.addressBlock-span').text()).toBe('first-address');
    expect(wrapper.find('.walletDiv-span').text()).toBe('first-balance');

    store.commit('changeBalance');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.addressBlock-span').text()).toBe('first-address');
    expect(wrapper.find('.walletDiv-span').text()).toBe('second-balance');

    store.commit('changeAddress');
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.addressBlock-span').text()).toBe('second-address');
    expect(wrapper.find('.walletDiv-span').text()).toBe('second-balance');
  });
});

it('copy button click', async () => {
  const store = createStore({
    state: {
      showModal: jest.fn(),
    },
  });

  wrapper = mount(WalletDiv, {
    global: {
      plugins: [store],
    },
  });

  expect(store.state.showModal).toBeCalledTimes(0);
  await wrapper.find('.addressBlock-button').trigger('click');
  expect(store.state.showModal).toBeCalledTimes(1);
});
