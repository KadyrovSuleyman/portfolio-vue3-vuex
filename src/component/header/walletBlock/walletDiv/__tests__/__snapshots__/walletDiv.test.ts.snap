// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectBlock renders 1`] = `
<div class="walletDiv">
  <div class="walletDiv-addressBlock"><span class="addressBlock-span"></span><button class="addressBlock-button addressBlock-button__icon_copy"></button></div><span class="walletDiv-span"> </span><button class="walletDiv-button walletDiv-button__icon_coin"><img></button><button class="walletDiv-button walletDiv-button__icon_chevron"></button>
</div>
`;
